kotlinx.fuzz.jazzer/src/main/kotlin/kotlinx/fuzz/jazzer/JazzerEngine.kt

```diff
@@ -1,60 +1,60 @@
-        Opt.instrumentationIncludes.setIfDefault(config.instrument)
-        Opt.customHookIncludes.setIfDefault(config.instrument)
-        Opt.customHookExcludes.setIfDefault(config.customHookExcludes)
+internal val KFuzzConfig.logsDir: Path
+    get() = workDir.resolve("logs")
 
-        AgentInstaller.install(Opt.hooks.get())
+@Suppress("unused", "SpellCheckingInspection")
+class JazzerEngine(private val config: KFuzzConfig) : KFuzzEngine {
+    private val jazzerConfig = JazzerConfig.fromSystemProperties()
 
-        FuzzTargetHolder.fuzzTarget = FuzzTargetHolder.FuzzTarget(
-            JazzerTarget::fuzzTargetOne.javaMethod,
-            LifecycleMethodsInvoker.noop(JazzerTarget),
-        )
+    override fun initialise() {
+        config.corpusDir.createDirectories()
+        config.logsDir.createDirectories()
     }
 
     @OptIn(ExperimentalPathApi::class)
     override fun runTarget(instance: Any, method: Method): Throwable? {
-        val libFuzzerArgs = mutableListOf("fake_argv0")
-        val corpusDir = createTempDirectory("jazzer-corpus")
-
-        libFuzzerArgs += corpusDir.toString()
-        libFuzzerArgs += "-max_total_time=${config.maxSingleTargetFuzzTime.inWholeSeconds}"
-        libFuzzerArgs += "-rss_limit_mb=${jazzerConfig.libFuzzerRssLimit}"
+        // spawn subprocess, redirect output to log and err files
+        val classpath = System.getProperty("java.class.path")
+        val javaCommand = System.getProperty("java.home") + "/bin/java"
+        val properties =
+            config.toPropertiesMap().map { (key, value) -> "-D$key=$value" }.toTypedArray()
+
+        val pb = ProcessBuilder(
+            javaCommand,
+            "-classpath", classpath,
+            *properties,
+            Launcher::class.qualifiedName!!,
+            method.declaringClass.name, method.name,
+        )
+        pb.redirectError(config.logsDir.resolve("${method.fullName}.err").toFile())
+        pb.redirectOutput(config.logsDir.resolve("${method.fullName}.log").toFile())
 
-        val atomicFinding = AtomicReference<Throwable>()
-        FuzzTargetRunner.registerFatalFindingHandlerForJUnit { finding ->
-            atomicFinding.set(finding)
+        val res = pb.start().waitFor()
+        if (res == 0) {
+            return null
         }
+        // TODO: read real exception
+        return Throwable("Jazzer subprocess returned with code $res")
+    }
 
-        JazzerTarget.reset(MethodHandles.lookup().unreflect(method), instance)
-        FuzzTargetRunner.startLibFuzzer(libFuzzerArgs)
-
-        corpusDir.deleteRecursively()
-
-        return atomicFinding.get()
+    override fun finishExecution() {
+        val statsDir = config.workDir.resolve("stats").createDirectories()
+        config.logsDir.listDirectoryEntries("*.err").forEach { file ->
+            val csvText = jazzerLogToCsv(file, config.maxSingleTargetFuzzTime)
+            statsDir.resolve("${file.nameWithoutExtension}.csv").writeText(csvText)
+        }
```

Comment:
Maybe collecting statistics should be a standalone function called in `finishExecution`? Every engine (most likely) will have this option, so why not add some `fun collectStatistics()`? And make `finishExecution` with default implementation (right now ` = collectStatistics()`) that can be overridden if needed

Reply:
I agree about extracting `collectStatistics`. However, I believe, adding it to `KFuzzEngine` interface is unnececarry. It enforces nothing, just one extra function.